### PR TITLE
feat: support GEMMA model

### PR DESCRIPTION
**Pull request by flaneur2020**
this PR adds the support of google's GEMMA model (CPU part only, i'd add the GPU part after the GPU side support quantization).

closes: #39

- [x] pass the tests in local machine
- [x] merge the main
- [x] cleanup hacks
- [x] refactor the weights loader

this PR also made these additional changes:

1. do not unquantize the embed table while loading, instead, unquantize it on `copy_from`. GEMMA's token table is too big.
2. add ModelArchitecture enum

### CODE REVIEW

File: crabml-core/src/backends/cpu/cpu_tensor.rs [hunk 1]
```diff
@@ -292,6 +292,14 @@ impl<'a> Tensor for CpuTensor<'a> {
         Ok(self)
     }
 
+    fn add_scalar_inplace(mut self, rhs: f32) -> Result<Self> {
+        let rhs = CpuTensor::new(vec![rhs], &[1], self.device())?;
+        let strider1 = self.strider().clone();
+        let strider2 = rhs.strider();
+        primitives::add_inplace(self.buf_mut(), rhs.buf(), &strider1, strider2)?;
+        Ok(self)
+    }
```
Comment (flaneur2020) [inline crabml-core/src/backends/cpu/cpu_tensor.rs:301]: this method is not in use, can be removed

File: crabml-core/src/backends/wgpu/wgpu_tensor.rs [hunk 2]
```diff
@@ -450,6 +450,66 @@ impl Tensor for WgpuTensor {
...
+                binding: 2,
+                resource: meta_buf.as_entire_binding(),
+            },
+        ];
+        let encoder = self
+            .device
+            .encode_pipeline_commnad("add_inplace", entries, (1, 1, 1));
+        self.device.queue.submit(Some(encoder.finish()));
+        Ok(self)
+    }
```
Comment (flaneur2020) [inline crabml-core/src/backends/wgpu/wgpu_tensor.rs:481]: can be removed

**Comment by flaneur2020** [comment]
Screenshot 2024-03-04 at 23 37 51

both have been supported

**Review by Xuanwo — APPROVED** [review]
Perfect!

**Pull request merged**